src/Persistence/IteratorQuery.php

```diff
@@ -0,0 +1,386 @@
+                    yield $id => array_intersect_key($row, $keys);
+                }
+            };
+        }
+    }
+
+    protected function initInsert(array $data): void
+    {
+        $this->fx = function (\Iterator $iterator) use ($data) {
+            return $this->persistence->setRawData($this->model, $data, $data[$this->model->id_field] ?? null);
+        };
+    }
+
+    protected function initUpdate(array $data): void
+    {
+        $this->fx = function (\Iterator $iterator) use ($data) {
+            foreach ($iterator as $id => $row) {
+                $this->persistence->setRawData($this->model, array_merge($row, $data), $data[$this->model->id_field] ?? $id);
+            }
+        };
+    }
+
+    protected function initDelete(): void
+    {
+        $this->fx = function (\Iterator $iterator) {
+            foreach ($iterator as $id => $row) {
+                $this->persistence->unsetRawData($this->model->table, $id);
+            }
+        };
+    }
+
+    /**
+     * Applies sorting on Iterator.
+     *
+     * @param array $fields
+     *
+     * @return $this
+     */
+    protected function initOrder(): void
+    {
+        if ($this->order) {
+            $data = $this->doGet();
+
+            // prepare arguments for array_multisort()
+            $args = [];
+            foreach ($this->order as [$field, $order]) {
+                $args[] = array_column($data, $field);
+                $args[] = $order === 'desc' ? SORT_DESC : SORT_ASC;
+            }
+            $args[] = &$data;
+
+            // call sorting
+            array_multisort(...$args);
+
+            // put data back in generator
+            $this->iterator = new \ArrayIterator(array_pop($args));
+        }
+    }
+
+    protected function initLimit(): void
+    {
+        if ($args = $this->getLimitArgs()) {
+            [$limit, $offset] = $args;
+
+            $this->iterator = new \LimitIterator($this->iterator, $offset, $limit);
+        }
+    }
+
+    /**
+     * Counts number of rows and replaces our generator with just a single number.
+     *
+     * @return $this
+     */
+    protected function initCount($alias = null): void
+    {
+        // @todo: kept for BC, inconstent results with SQL count!
+        $this->initLimit();
+
+        $alias = $alias ?? 'count';
+
+        $this->iterator = new \ArrayIterator([[$alias => iterator_count($this->iterator)]]);
+    }
+
+    /**
+     * Checks if iterator has any rows.
+     *
+     * @return $this
+     */
+    protected function initExists(): void
+    {
+        $this->iterator = new \ArrayIterator([[$this->iterator->valid() ? 1 : 0]]);
+    }
+
+    protected function initField($fieldName, string $alias = null): void
+    {
+        if (!$fieldName) {
+            throw new Exception('Field query requires field name');
+        }
+
+        // get first record
+        if ($row = $this->persistence->query($this->model)->select([$fieldName])->getRow()) {
```

Comment:
I wonder why this https://github.com/atk4/data/pull/861 bug passed test in this refactoring...